test/test_selectors.py

```diff
@@ -681,24 +689,93 @@ def test_above_fd_setsize(self):
         self.assertEqual(limit_nofile // 2, len(s.select()))
 
 
+@skipUnless(selectors.HAS_SELECT, "Platform doesn't have a selector.")
+class TestUniqueSelectScenarios(BaseSelectorTestCase):
+    def test_select_module_patched_after_import(self):
+        # This test is to make sure that after import time
+        # calling DefaultSelector() will still give a good
+        # return value. This issue is caused by gevent, eventlet.
+
+        # Now remove all selectors except `select.select`.
+        self.setup_select_module('select')
+
+        # Make sure that the selector returned only uses the selector available.
+        selector = self.make_selector()
+        self.assertIsInstance(selector, selectors.SelectSelector)
+
+    @skipUnless(hasattr(errno, 'ENOSYS'), 'Platform does not have errno.ENOSYS.')
+    def test_select_module_defines_does_not_implement_poll(self):
+        # This test is to make sure that if a platform defines
+        # a selector as being available but does not actually
+        # implement it (kennethreitz/requests#3906) does not
+        # make DefaultSelector() fail.
+
+        # Reset the _DEFAULT_SELECTOR value as if using for the first time.
+        selectors._DEFAULT_SELECTOR = None
+
+        had_poll = hasattr(select, 'poll')
+
+        # Remove all selectors except `select.select`.
+        self.setup_select_module('select')
+
+        # Now we're going to patch in a bad `poll`.
+        class BadPoll(object):
+            def poll(self, timeout):
+                raise OSError(errno.ENOSYS)
+
+        select.poll = BadPoll
+        if not had_poll:
+            self.addCleanup(delattr, select, 'poll')
+
+        selector = self.make_selector()
+        self.assertIsInstance(selector, selectors.SelectSelector)
+
+    @skipUnless(hasattr(errno, 'ENOSYS'), 'Platform does not have errno.ENOSYS.')
+    def test_select_module_defines_does_not_implement_epoll(self):
+        # Same as above test except with `select.epoll`.
+
+        # Reset the _DEFAULT_SELECTOR value as if using for the first time.
+        selectors._DEFAULT_SELECTOR = None
+
+        had_epoll = hasattr(select, 'epoll')
+
+        # Remove all selectors except `select.select`.
+        self.setup_select_module('select')
+
+        # Now we're going to patch in a bad `epoll`.
+        def bad_epoll(*args, **kwargs):
+            raise OSError(errno.ENOSYS)
+
+        select.epoll = bad_epoll
+        if not had_epoll:
+            self.addCleanup(delattr, select, 'epoll')
```

Comment:
So this patching and unpatching is a really common pattern in these tests: it might be good to factor it out to a helper method like `patch_selector_implementation('epoll', bad_epoll)`